crates/ide_assists/src/handlers/extract_module.rs

```diff
@@ -0,0 +1,1394 @@
+                if selection_range.contains_range(item.syntax().text_range()) {
+                    return Some(item);
+                }
+                return None;
+            }
+            None
+        })
+        .collect();
+
+    return Some(Module {
+        text_range: selection_range,
+        name: "modname".to_string(),
+        body_items,
+        leading_ws: "".to_string(),
+    });
+}
+
+impl Module {
+    fn update_refs(
+        &self,
+        covering_ele: &SyntaxNode,
+        ctx: &AssistContext,
+        builder: &mut AssistBuilder,
+    ) -> Option<Vec<ast::Item>> {
+        let (refs, record_fields) = self.get_refs(covering_ele, ctx);
+
+        for (file_id, references) in refs {
+            let source_file = ctx.sema.parse(file_id);
+
+            builder.edit_file(file_id);
+
+            for usage in references {
+                self.process_usage(&source_file, builder, usage);
+            }
+        }
+
+        let (body_items, mut replacements, record_field_parents, impls) =
+            get_replacements(self.body_items.clone(), false);
+
+        let impl_items = impls.into_iter().fold(Vec::new(), |mut impl_items, x| {
+            let this_impl_items =
+                x.syntax().descendants().fold(Vec::new(), |mut this_impl_items, x| {
+                    if let Some(item) = ast::Item::cast(x.clone()) {
+                        this_impl_items.push(item);
+                    }
+                    return this_impl_items;
+                });
+
+            impl_items.append(&mut this_impl_items.clone());
+            return impl_items;
+        });
+
+        let (_, mut impl_item_replacements, _, _) = get_replacements(impl_items.clone(), true);
+
+        replacements.append(&mut impl_item_replacements);
+
+        record_field_parents.into_iter().for_each(|x| {
+            x.1.descendants().filter_map(|x| ast::RecordField::cast(x)).for_each(|desc| {
+                let is_record_field_present = record_fields
+                    .clone()
+                    .into_iter()
+                    .find(|x| x.to_string() == desc.to_string())
+                    .is_some();
+                if is_record_field_present {
+                    replacements.push((desc.visibility().clone(), desc.syntax().clone()));
+                }
+            });
+        });
+
+        replacements.into_iter().for_each(|(vis, syntax)| {
+            add_change_vis(vis, syntax.first_child_or_token());
+        });
+
+        Some(body_items)
+    }
+
+    fn get_refs(
+        &self,
+        ele: &SyntaxNode,
+        ctx: &AssistContext,
+    ) -> (HashMap<FileId, Vec<FileReference>>, Vec<SyntaxNode>) {
+        let mut record_fields = Vec::new();
+        let mut refs: HashMap<FileId, Vec<FileReference>> = HashMap::new();
+
+        let selection_range = ctx.frange.range;
+
+        for child in ele.children() {
+            if selection_range.contains_range(child.text_range()) {
+                match_ast! {
```

Comment:
This is already the information we have saved in `Module::body_items` if I am not mistaken, so instead we can just go through that vec instead of filtering through the node children here again no? That should simplify this somewhat